discretenet/problems/gisp/gisp.py

```diff
@@ -0,0 +1,216 @@
+        self.E2 = set()
+
+        if self.graph_instance is None:
+            # Generate random graph
+            num_nodes = random.randint(self.min_n, self.max_n)
+            self.graph = nx.erdos_renyi_graph(
+                n=num_nodes, p=self.er_prob, seed=self.random_seed
+            )
+            self.name = "er_n=%d_m=%d_p=%.2f_%s_setparam=%.2f_alpha=%.2f" % (
+                num_nodes,
+                nx.number_of_edges(self.graph),
+                self.er_prob,
+                self.which_set,
+                self.set_param,
+                self.alpha,
+            )
+        else:
+            self.dimacs_to_nx()
+            self.name = "%s_%s_%g_%g" % (
+                self.graph_instance,
+                self.which_set,
+                self.alpha,
+                self.set_param,
+            )
+
+    def dimacs_to_nx(self):
+        g = nx.Graph()
+        with open_text("discretenet.problems.gisp.graphs", self.graph_instance) as f:
+            for line in f:
+                arr = line.split()
+                if line[0] == "e":
+                    g.add_edge(int(arr[1]), int(arr[2]))
+        self.graph = g
+
+    def problem(self) -> Problem:
+        """
+        A :class:`discretenet.problem.Problem` instance generated by this generator.
+        Must be set when creating a concrete generator. ::
+            class MyGenerator(Generator):
+                problem = MyProblem
+                ...
+        """
+        return self.problem
+
+    def generate(self):
+        """
+        Generate and return a single :class:`discretenet.problem.Problem` instance
+        Should sample for parameters from the defined parameter generator callables
+        (from the ``__init__()`` method), instantiate an instance with those parameters,
+        and return it.
+        :return: An initialized concrete ``Problem`` instance
+        """
+
+        # Generate node revenues and edge costs
+        self.generate_revs_costs()
+
+        # Generate the set of removable edges
+        self.generate_E2()
+
+        # Construct Problem
+        problem = GISPProblem(self.graph, self.E2, self.name + "_%d" % self.random_seed)
+
+        # Update seed
+        self.set_seed(self.random_seed + 1)
+
+        return problem
+
+    def generate_revs_costs(self):
+        if self.which_set == "SET1":
+            for node in self.graph.nodes():
+                self.graph.nodes[node]["revenue"] = random.randint(1, 100)
+            for u, v, edge in self.graph.edges(data=True):
+                edge["cost"] = (
+                    self.graph.node[u]["revenue"] + self.graph.node[v]["revenue"]
+                ) / float(self.set_param)
+        elif self.which_set == "SET2":
+            for node in self.graph.nodes():
+                self.graph.nodes[node]["revenue"] = float(self.set_param)
+            for u, v, edge in self.graph.edges(data=True):
+                edge["cost"] = 1.0
+
+    def generate_E2(self):
```

Comment:
Because `self.E2` isn't cleared between each generate call, it's getting bigger each time and each new generated instance will depend on the previous one. I'm not sure about the problem specifically, but is that intentional? My intuition is that each call to `generate()` should be independent (`generate_revs_cost` has this issue too)

Reply:
Thanks for pointing this out, I updated the code so that E2 is redefined as a set every time `generate()` is called

Reply:
Looks better! 

For both this an `generate_revs_cost()`, could you make them just return the relevant thing (set / graph) rather than setting `self.E2` and modifying `self.graph`? I assume you'll need to pass the graph that `generate_revs_cost()` returns into `generate_E2` to do that, but it will make the code a bit easier to follow and debug.

Reply:
Updated